test/unit/asrouter/CFRPageActions.test.js

```diff
@@ -0,0 +1,641 @@
+        clock.tick(delay + 1);
+        assert.equal(pageAction.urlbar.getAttribute("cfr-recommendation-state"), "expanded");
+      });
+    });
+
+    describe("#_collapse", () => {
+      beforeEach(() => {
+        pageAction._clearScheduledStateChanges();
+        pageAction.urlbar.removeAttribute("cfr-recommendation-state");
+      });
+      it("without a delay, should clear other state changes and set the state to collapsed only if it's already expanded", () => {
+        sandbox.spy(pageAction, "_clearScheduledStateChanges");
+        pageAction._collapse();
+        assert.calledOnce(pageAction._clearScheduledStateChanges);
+        assert.isNull(pageAction.urlbar.getAttribute("cfr-recommendation-state"));
+        pageAction.urlbar.setAttribute("cfr-recommendation-state", "expanded");
+        pageAction._collapse();
+        assert.equal(pageAction.urlbar.getAttribute("cfr-recommendation-state"), "collapsed");
+      });
+      it("with a delay, should set the collapsed state after the correct amount of time", () => {
+        const delay = 1234;
+        pageAction._collapse(delay);
+        clock.tick(delay + 1);
+        // The state was _not_ "expanded" and so should not have been set to "collapsed"
+        assert.isNull(pageAction.urlbar.getAttribute("cfr-recommendation-state"));
+
+        pageAction._expand();
+        pageAction._collapse(delay);
+        // We expect that a collapse has been scheduled
+        assert.lengthOf(pageAction.stateTransitionTimeoutIDs, 1);
+        clock.tick(delay + 1);
+        // This time it was "expanded" so should now (after the delay) be "collapsed"
+        assert.equal(pageAction.urlbar.getAttribute("cfr-recommendation-state"), "collapsed");
+      });
+    });
+
+    describe("#_clearScheduledStateChanges", () => {
+      it("should call .clearTimeout on all stored timeoutIDs", () => {
+        pageAction.stateTransitionTimeoutIDs = [42, 73, 1997];
+        sandbox.spy(pageAction.window, "clearTimeout");
+        pageAction._clearScheduledStateChanges();
+        assert.calledThrice(pageAction.window.clearTimeout);
+        assert.calledWith(pageAction.window.clearTimeout, 42);
+        assert.calledWith(pageAction.window.clearTimeout, 73);
+        assert.calledWith(pageAction.window.clearTimeout, 1997);
+      });
+    });
+
+    describe("#_popupStateChange", () => {
+      it("should collapse and remove the notification on 'dismissed' or 'removed'", () => {
+        pageAction._expand();
+        const fakeNotification = {};
+
+        pageAction.currentNotification = fakeNotification;
+        pageAction._popupStateChange("dismissed");
+        assert.equal(pageAction.urlbar.getAttribute("cfr-recommendation-state"), "collapsed");
+        assert.calledOnce(global.PopupNotifications.remove);
+        assert.calledWith(global.PopupNotifications.remove, fakeNotification);
+
+        pageAction.currentNotification = fakeNotification;
+        sandbox.resetHistory();
```

Comment:
can you split this into two tests instead of the reset? Should make it easier to debug in the future.